### ISSUE TITLE
implement aggregate transformer

### ISSUE THREAD

**Issue by 100pah** [apache/echarts#15306]
At present the there is only a simplest aggregate transformer in [echarts-simple-transform/aggregate](https://github.com/100pah/echarts-simple-transform/blob/main/src/aggregate.ts). We should better implement a built-in aggregate transformer in echarts.

## About the API
Think of whether the API of [echarts-simple-transform/aggregate](https://github.com/100pah/echarts-simple-transform/blob/main/src/aggregate.ts) appropriate.

## About group by
In [echarts-simple-transform/aggregate](https://github.com/100pah/echarts-simple-transform/blob/main/src/aggregate.ts) only on dimension can be set in group by. Do we need to support multiple dimension group by?

## About Methods

Some `methods` of aggregate already roughly implemented in [echarts-simple-transform/aggregate](https://github.com/100pah/echarts-simple-transform/blob/main/src/aggregate.ts):
+ `sum`
+ `count`
+ `first`
+ `average`
+ `Q1`
+ `Q2`/`median`
+ `Q3`
+ `min`
+ `max`

We need to make sure they are implemented correctly.

Other methods we may implement:
+ `distinct`: The count of distinct dimension values.
+ `variance`: The sample variance of dimension values.
+ `variancep`: The population variance of dimension values.
+ `stdev`: The sample standard deviation of dimension values.
+ `stdevp`: The population standard deviation of dimension values.
+ `stderr`: The standard error of dimension values.
+ `product`: The product of dimension values.
+ `valid`: The count of dimension values that are not `null`, `undefined`, `''`, `NaN`, `'-'`, or determined by user defined validation function.
+ ... anything else useful?

Some references: [vega aggregate](https://vega.github.io/vega/docs/transforms/aggregate/).

## About math calculation precision

We all known that the floating precision issue (like `0.1 + 0.2`). It might bring trouble in some scenarios like:
+ The math result is needed to be displayed in label.
+ The sum should equal to 100%.
+ Multiple by 100, 1000, ...

See: [IEEE 754](https://en.wikipedia.org/wiki/IEEE_floating_point), <http://0.30000000000000004.com/>, <https://github.com/camsong/blog/issues/9>, <https://javascript.info/number#imprecise-calculations>, ... (lots of related references)

Some cases:
```js
0.1 + 0.2 === 0.30000000000000004
2.3 - 0.3 === 1.9999999999999998
33643.2 - 15918.3 === 17724.899999999998
146.39 - 62.83 === 83.55999999999999
1.09 - 0.13 === 0.9600000000000001
0.1 * 0.2 === 0.020000000000000004
16.08 * 100 === 1607.9999999999998
146.39 * 100 === 14638.999999999998
9999999999999999 === 10000000000000000
-11.699999999999986 === -11.699999999999987
```

We should also take care the `toFixed bug` like:
```js
1.005.toFixed(2) === '1.00' // rather than '1.01'
0.1.toFixed(20) === '0.10000000000000000555'
```

**Solution:**
+ A. Do not care about it until some users required.
+ B. Try to resolve it:
    + For arithmetic: use [addSafe](https://github.com/apache/echarts/blob/5.1.2/src/util/number.ts#L286) we already have and implement `multipleSafe` in appropriate place.
    + For comparison of two numbers, use:
    ```js
    const EPSILON = (function () { // Solve compatibility issues
        return Number.EPSILON ? Number.EPSILON : Math.pow(2, -52);
    })();
    function compare(a, b){
        return Math.abs(a - b) < EPSILON;
    }
    ```

## About big number

Do not support.

## Test

Could find some test cases in some other math libraries.

**Comment by maneetgoyal** [comment]
Would be great to have this feature 💯

**Comment by sumitnegi7** [comment]
Seems like a pretty useful feature

**Comment by saikat-samanta** [comment]
This feature seems really helpful. 👍

**Comment by Prit-sangani** [comment]
Yes, this feature is very helpful. Would be great to have this feature.

**Comment by VirajBhatt9** [comment]
It is a very great feature. It would be very helpful to have this feature.

**Comment by swapnil0545** [comment]
Any update on merging this?

**Comment by MrZhouZh** [comment]
<img width="319" alt="image" src="https://github.com/apache/echarts/assets/24643748/5f9cc5e8-2181-4590-acf9-216ab89b7686">

This feature that i really want it. cannot wait

**Comment by namannitr** [comment]
Following up on “we need to make sure they [aggregate methods] are implemented correctly” before a core migration.

The experimental implementation in [echarts-simple-transform](https://github.com/100pah/echarts-simple-transform) still has two SUM correctness issues that should not be copied into core as-is:

1. **Published `dist` still seeds `SUM` with `0`**, so a group with one row sums to `0` and every other group **drops the first row**. The TypeScript source was fixed in 100pah/echarts-simple-transform#3 (2021) but never rebuilt/released; jsDelivr still serves the broken bundle. See 100pah/echarts-simple-transform#1.
2. **String-valued cells concatenate** (`'400' + '100' === '400100'`). Typical `objectRows` SQL/JSON sources hit this. See 100pah/echarts-simple-transform#6.

Fix + tests + rebuilt `dist`: https://github.com/100pah/echarts-simple-transform/pull/9

Draft core migration #16903 should take the same two behaviors: seed `SUM` from the first in-group numeric value (or skip NaN), and coerce with `Number` / the existing dataset parsers rather than `+` on `OptionDataValue`.

Separately, aggregate alone does not solve long-format → multiple series (that needs a pivot / groupBy reshape). Opened #21742 for that, linked from #16083.
